go/parquet/metadata/statistics.go

```diff
@@ -0,0 +1,517 @@
+func (Int32Statistics) equal(a, b int32) bool                 { return a == b }
+func (Int64Statistics) equal(a, b int64) bool                 { return a == b }
+func (Float32Statistics) equal(a, b float32) bool             { return a == b }
+func (Float64Statistics) equal(a, b float64) bool             { return a == b }
+func (Int96Statistics) equal(a, b parquet.Int96) bool         { return bytes.Equal(a[:], b[:]) }
+func (ByteArrayStatistics) equal(a, b parquet.ByteArray) bool { return bytes.Equal(a, b) }
+func (FixedLenByteArrayStatistics) equal(a, b parquet.FixedLenByteArray) bool {
+	return bytes.Equal(a, b)
+}
+
+func (BooleanStatistics) less(a, b bool) bool {
+	return !a && b
+}
+
+func (s *Int32Statistics) less(a, b int32) bool {
+	if s.order == schema.SortUNSIGNED {
+		return uint32(a) < uint32(b)
+	}
+	return a < b
+}
+
+func (s *Int64Statistics) less(a, b int64) bool {
+	if s.order == schema.SortUNSIGNED {
+		return uint64(a) < uint64(b)
+	}
+	return a < b
+}
+func (Float32Statistics) less(a, b float32) bool { return a < b }
+func (Float64Statistics) less(a, b float64) bool { return a < b }
+func (s *Int96Statistics) less(a, b parquet.Int96) bool {
+	i96a := arrow.Uint32Traits.CastFromBytes(a[:])
+	i96b := arrow.Uint32Traits.CastFromBytes(b[:])
+	if i96a[2] != i96b[2] {
+		// only the msb bit is by signed comparison
+		if s.order == schema.SortSIGNED {
+			return int32(i96a[2]) < int32(i96b[2])
+		}
+		return i96a[2] < i96b[2]
+	} else if i96a[1] != i96b[1] {
+		return i96a[1] < i96b[1]
+	}
+	return i96a[0] < i96b[0]
+}
+
+func (s *ByteArrayStatistics) less(a, b parquet.ByteArray) bool {
+	if s.order == schema.SortUNSIGNED {
+		return bytes.Compare(a, b) == -1
+	}
+
+	sa := *(*[]int8)(unsafe.Pointer(&a))
+	sb := *(*[]int8)(unsafe.Pointer(&b))
+	i := 0
+	for ; i < len(sa) && i < len(sb); i++ {
+		if sa[i] < sb[i] {
+			return true
+		}
+		if sb[i] < sa[i] {
+			return false
+		}
+	}
+	return i == len(sa) && i != len(sb)
+}
+
+func (s *FixedLenByteArrayStatistics) less(a, b parquet.FixedLenByteArray) bool {
+	if s.order == schema.SortUNSIGNED {
+		return bytes.Compare(a, b) == -1
+	}
+
+	sa := *(*[]int8)(unsafe.Pointer(&a))
+	sb := *(*[]int8)(unsafe.Pointer(&b))
+	i := 0
+	for ; i < len(sa) && i < len(sb); i++ {
+		if sa[i] < sb[i] {
+			return true
+		}
+		if sb[i] < sa[i] {
+			return false
+		}
+	}
+	return i == len(sa) && i != len(sb)
+}
+
+func (BooleanStatistics) cleanStat(minMax minmaxPairBoolean) *minmaxPairBoolean { return &minMax }
+func (Int32Statistics) cleanStat(minMax minmaxPairInt32) *minmaxPairInt32       { return &minMax }
+func (Int64Statistics) cleanStat(minMax minmaxPairInt64) *minmaxPairInt64       { return &minMax }
+func (Int96Statistics) cleanStat(minMax minmaxPairInt96) *minmaxPairInt96       { return &minMax }
+
+// in the case of floating point types, the following rules are applied as per parquet-mr:
+// - if any of min/max is NaN, return nothing
```

Comment:
i don't think it is done consistently across all implementations and I think there is an open JIRA to try to standardize it.

Reply:
For this, once again, i copied it from the C++ implementation which had a comment that referred to the parquet-mr impl, which is why i copied the comment over too. Should i mention that it's not necessarily standardized across the impls yet?

Reply:
Yeah, we should reference https://issues.apache.org/jira/browse/PARQUET-1222

Reply:
Cool, I'll add a comment and reference that.

Reply:
added comment.